### PR TITLE
Add API docs

### PR DESCRIPTION
**Pull request by ryantwolf**
## Description

Adds the skeleton of the API docs that will be rendered on the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/datacuration/index.html). Some existing docstrings are outdated or incomplete, but those will be addressed in another PR. I did modify a few though to see how they would render on the autodoc output, so that's why a few random files have been changed.

## Usage

N/A
## Checklist

- [x] I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- [x] New or Existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: nemo_curator/modules/filter.py [hunk 1]
```diff
@@ -26,19 +29,45 @@
 
 
 class Score:
-    def __init__(self, score_fn, score_field, text_field="text", score_type=None):
+    """
+    The module responsible for adding metadata to records based statistics about the text.
```
Comment (sarahyurick) [inline nemo_curator/modules/filter.py:33]:
```suggestion
    The module responsible for adding metadata to records based on statistics about the text.
```
?

File: nemo_curator/modules/filter.py [hunk 2]
```diff
@@ -86,26 +134,49 @@ def __call__(self, dataset):
...
+        text_field: str = "text",
+        score_field: Optional[str] = None,
+        score_type: Union[type, str] = None,
+        invert: bool = False,
     ):
         """
+        Constructs a ScoreFilter module.
+
         Args:
+          filter_obj (DocumentFilter): The score function that takes in a document string and outputs a score for the document
```
Comment (sarahyurick) [inline nemo_curator/modules/filter.py:158]:
```suggestion
          filter_obj (DocumentFilter): The score function that takes in a document string and outputs a score for the document.
```

File: nemo_curator/modules/filter.py [hunk 3]
```diff
@@ -26,19 +29,45 @@
...
+        score_type: Union[type, str] = None,
+    ):
         """
+        Constructs a Score module.
+
         Args:
-          score_fn: The score function that takes in a document string and outputs a score for the document
-          score_field: The field the score will be stored in.
-          text_field: The field the documents will be read from
+          score_fn (Callable): The score function that takes in a document string and outputs a score for the document
```
Comment (sarahyurick) [inline nemo_curator/modules/filter.py:51]:
```suggestion
          score_fn (Callable): The score function that takes in a document string and outputs a score for the document.
```

File: nemo_curator/modules/filter.py [hunk 4]
```diff
@@ -58,18 +87,37 @@ def __call__(self, dataset):
...
+
+    def __init__(self, filter_fn: Callable, filter_field: str, invert: bool = False):
         """
+        Constructs a Filter module
+
         Args:
-          filter_fn: A function that returns True if the document is to be kept
-          filter_field: The field(s) to be passed into the filter function.
-          invert: Whether to invert the filter condition
+          filter_fn (Callable): A function that returns True if the document is to be kept
```
Comment (sarahyurick) [inline nemo_curator/modules/filter.py:103]:
```suggestion
          filter_fn (Callable): A function that returns True if the document is to be kept.
```

File: nemo_curator/modules/filter.py [hunk 5]
```diff
@@ -58,18 +87,37 @@ def __call__(self, dataset):
...
+          filter_field (str): The field(s) to be passed into the filter function.
+          invert (bool): Whether to invert the filter condition
```
Comment (sarahyurick) [inline nemo_curator/modules/filter.py:105]:
```suggestion
          invert (bool): Whether to invert the filter condition.
```

File: docs/user-guide/index.rst [hunk 6]
```diff
@@ -42,6 +42,8 @@
 `Tutorials <https://github.com/NVIDIA/NeMo-Curator/tree/main/tutorials>`__
    To get started, you can explore the NeMo Curator GitHub repository and follow the available tutorials and notebooks. These resources cover various aspects of data curation, including training from scratch and Parameter-Efficient Fine-Tuning (PEFT).
 
+:ref:`API Docs <data-curator-api>`
+   API Docuemntation for all the modules in NeMo Curator
```
Comment (sarahyurick) [inline docs/user-guide/index.rst:46]:
```suggestion
   API Documentation for all the modules in NeMo Curator
```

File: docs/user-guide/api/index.rst [hunk 7]
```diff
@@ -0,0 +1,21 @@
+.. _data-curator-api:
+
+======================================
+API Reference
+======================================
+
+.. toctree::
+   :maxdepth: 4
+   :titlesonly:
```
Comment (sarahyurick) [inline docs/user-guide/api/index.rst:9]: What is this for?

**Review by sarahyurick — COMMENTED** [review]
LGTM, just added a bunch of nits.

File: docs/user-guide/api/index.rst (hunk 7 above)
Comment (ryantwolf) [reply docs/user-guide/api/index.rst:9]: It's so that all the options show up in the sidebar navigation menu.

**Comment by ryantwolf** [comment]
Thanks @sarahyurick. I think I addressed all your comments.

**Review by sarahyurick — APPROVED** [review]
LGTM, thanks!

**Pull request merged**